Core: Ignore split offsets array when split offset is past file length

amogh-jahagirdar opened the pull request:
Follow up from a miss in the fix done in https://github.com/apache/iceberg/pull/8860. There's an additional splitOffsetArray method which gets used here https://github.com/apache/iceberg/blob/aa891acf20040d15e7ca59dc503adb3c1e4325b8/core/src/main/java/org/apache/iceberg/BaseContentScanTask.java#L134 . We would want to surface null for this array in case we know the split offsets are corrupted and invalid (stemming from an issue in 1.4.0). The split offsets are invalid when the final offset is beyond the end of the file.

Alternatively, we can undo the optimization done here: https://github.com/apache/iceberg/pull/8336#discussion_r1296190234 and then the original fix in #8860 would be sufficient, and be simpler. Since the optimization seemed impactful and there was community interest in it, this fix retains it and currently just fixes both cases explicitly.

rdblue commented:
@amogh-jahagirdar, maybe this time we should create a test to validate `FileScanTask.split` when there are bad split offsets? I think that would have caught this in the last PR.

amogh-jahagirdar commented:
> @amogh-jahagirdar, maybe this time we should create a test to validate FileScanTask.split when there are bad split offsets? I think that would have caught this in the last PR.

Done, added a new test which will exercise both `splitOffsetArray` and `splitOffsets`. This will also validate the number of tasks per file

nastra commented on core/src/test/java/org/apache/iceberg/TestSplitPlanning.java:
```diff
@@ -216,6 +217,34 @@ public void testSplitPlanningWithOffsets() {
+            .option(TableProperties.SPLIT_OPEN_FILE_COST, String.valueOf(0))
+            .option(TableProperties.SPLIT_SIZE, String.valueOf(4L * 1024 * 1024));
+
+    Assert.assertEquals(
```
minor: for newly added test code it would be better to use AssertJ-style assertions to make a future migration away from JUnit4 easier

amogh-jahagirdar replied:
Done, this test is no longer applicable but for the new test it's using AssertJ

nastra commented on core/src/test/java/org/apache/iceberg/TestSplitPlanning.java:
```diff
@@ -216,6 +217,34 @@ public void testSplitPlanningWithOffsets() {
+      Collection<FileScanTask> files = task.files();
+      for (FileScanTask fileScanTask : files) {
+        DataFile dataFile = fileScanTask.file();
+        Assert.assertNull(
```
same as above

nastra commented on core/src/test/java/org/apache/iceberg/TestSplitPlanning.java:
```diff
@@ -216,6 +217,34 @@ public void testSplitPlanningWithOffsets() {
   }
 
+  @Test
+  public void testSplitPlanningWithCorruptedOffsets() throws IOException {
```
unfortunately that test passes for me locally without the new fix in `BaseFile`

amogh-jahagirdar replied:
Discussed offline, this test exercises the path but has poor test hardness since the offsets end up being null anyways.
For proper test hardness we can go through TableScanUtil.planTaskGroups which will call FileScanTask.split and exercise the split offset array path. The new test will fail without the fix since the number of tasks will be different (without the fix, there will be 2 tasks, 1 per split offset but the last task is a bad task. With the fix, the split size of 1 byte will be used for the 10 byte file, and there will be 10 tasks)
